qualang_tools/optimal_weights/StateDiscriminator.py

```diff
@@ -0,0 +1,256 @@
+                I_ = I_res[self.seq0 == i]
+                Q_ = Q_res[self.seq0 == i]
+                plt.plot(I_, Q_, '.', label=f'state {i}')
+                plt.axis('equal')
+            plt.xlabel('I')
+            plt.ylabel('Q')
+            plt.legend()
+
+            plt.figure()
+            for i in range(self.num_of_states):
+                plt.subplot(self.num_of_states, 1, i + 1)
+                plt.plot(np.real(weights[i, :]))
+                plt.plot(np.imag(weights[i, :]))
+
+            plt.figure()
+            for i in range(self.num_of_states):
+                plt.plot(np.real(weights[i, :]), np.imag(weights[i, :]))
+                plt.axis('equal')
+
+    def _add_iw_to_all_pulses(self, iw):
+        for pulse in self.config['pulses'].values():
+            if 'integration_weights' not in pulse:
+                pulse['integration_weights'] = {}
+            pulse['integration_weights'][iw] = iw
+
+    def _update_config(self):
+        weights = self.saved_data['weights']
+        for i in range(self.num_of_states):
+            self.config['integration_weights'][f'state_{i}_in1_{self.rr_qe}'] = {
+                'cosine': np.real(weights[i, :]).tolist(),
+                'sine': (-np.imag(weights[i, :])).tolist()
+            }
+            self._add_iw_to_all_pulses(f'state_{i}_in1_{self.rr_qe}')
+            self.config['integration_weights'][f'state_{i}_in2_{self.rr_qe}'] = {
+                'cosine': np.imag(weights[i, :]).tolist(),
+                'sine': np.real(weights[i, :]).tolist()
+            }
+            self._add_iw_to_all_pulses(f'state_{i}_in2_{self.rr_qe}')
+            if self.update_tof or self.finish_train == 1:
+                self.config['elements'][self.rr_qe]['time_of_flight'] = self.config['elements'][self.rr_qe][
+                                                                            'time_of_flight'] - \
+                                                                        self.config['elements'][self.rr_qe]['smearing']
+
+    def measure_state(self, pulse, out1, out2, res, adc=None):
+        """
+        This procedure generates a macro of QUA commands for measuring the readout resonator and discriminating between
+        the states of the qubit its states.
+        :param pulse: A string with the readout pulse name.
+        :param out1: A string with the name first output of the readout resonator (corresponding to the real part of the
+         complex IN(t_int) signal).
+        :param out2: A string with the name second output of the readout resonator (corresponding to the imaginary part
+        of the complex IN(t_int) signal).
+        :param res: An integer QUA variable that will receive the discrimination result (0,1 or 2)
+        :param adc: (optional) the stream variable which the raw ADC data will be saved and will appear in result
+        analysis scope.
+        """
+        bias = self.saved_data['bias']
+        # currently it allows only 3 states
```

Comment:
I think the original idea is that it can be generalized to beyond so that is why it had that name. It also has the different functions that the TwoStateDisc inherets to do its job

Reply:
yes, but as of now I have simplified it to TwoState because we have not done it yet for further